docs/kernel_finder.rst

```diff
@@ -0,0 +1,183 @@
+
+
+
+The user location takes priority over the system locations, and the case of the
+names is ignored, so selecting kernels works the same way whether or not the
+filesystem is case sensitive.
+
+Since kernel names, and their :ref:`provider ids <provider_id>`, show up in URLs and other places,
+a kernelspec is required to have a simple name, only containing ASCII letters, ASCII numbers, and the simple separators: ``-`` hyphen, ``.`` period, ``_`` underscore.
+
+Other locations may also be searched if the :envvar:`JUPYTER_PATH` environment
+variable is set.
+
+For IPython kernels, three types of files are presently used:
+``kernel.json``, ``kernel.js``, and logo image files. However, different Kernel Providers
+can support other files and directories within the kernel directory or may not even
+use a directory for their kernel discovery model.  That said, for kernels prior
+to Kernel Providers or those discovered by instances of class
+:class:`.KernelSpecProvider`, the most important
+file is **kernel.json**. This file consists of a JSON-serialized dictionary
+that adheres to the :ref:`kernel specification format <kernelspec_format>`.
+
+
+For example, the kernel.json file for the IPython kernel looks like this::
+
+    {
+     "argv": ["python3", "-m", "IPython.kernel",
+              "-f", "{connection_file}"],
+     "display_name": "Python 3",
+     "language": "python"
+    }
+
+
+.. _kernelspec_format:
+
+Kernel Specification Format
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The information contained in each entry returned from a Kernel Provider's
+:meth:`find_kernels() <.KernelProviderBase.find_kernels>` method consists of a
+JSON serialised dictionary containing the following keys and values:
+
+- **display_name**: The kernel's name as it should be displayed in the UI.
+  Unlike the kernel name used in the API, this can contain arbitrary unicode
+  characters.  This value should be provided by all kernel providers.
+- **language**: The name of the language of the kernel.
+  When loading notebooks, if no matching kernelspec key (may differ across machines)
+  is found, a kernel with a matching `language` will be used.
+  This allows a notebook written on any Python or Julia kernel to be properly
+  associated with the user's Python or Julia kernel, even if they aren't listed
+  under the same name as the author's. This value should be provided by all kernel providers.
+- **argv**: (optional): A list of command line arguments used to start the kernel. For
+  instances of class :class:`KernelSpecProvider <jupyter_kernel_mgmt.discovery.KernelSpecProvider>` the text
+  ``{connection_file}`` in any argument will be replaced with the path to the
+  connection file.  However, subclasses of :class:`KernelSpecProvider <jupyter_kernel_mgmt.discovery.KernelSpecProvider>`
+  may choose to provide different substitutions, especially if they don't use a connection file.
+- **interrupt_mode** (optional): May be either ``signal`` or ``message`` and
```

Comment:
Likewise here: kernel specs need to tell the provider if they don't want the default mode, but that doesn't need to be exposed through the kernel type discovery mechanism.

The dicts from `find_kernels` are (IMO) primarily there to help the user or the system select an appropriate kernel to launch.